core/src/main/java/com/datastax/oss/driver/internal/core/cql/CqlPrepareAsyncProcessor.java

```diff
@@ -15,34 +15,78 @@
  */
 package com.datastax.oss.driver.internal.core.cql;
 
+import com.datastax.oss.driver.api.core.cql.ColumnDefinition;
 import com.datastax.oss.driver.api.core.cql.PrepareRequest;
 import com.datastax.oss.driver.api.core.cql.PreparedStatement;
 import com.datastax.oss.driver.api.core.session.Request;
+import com.datastax.oss.driver.api.core.type.UserDefinedType;
 import com.datastax.oss.driver.api.core.type.reflect.GenericType;
+import com.datastax.oss.driver.internal.core.context.DefaultDriverContext;
 import com.datastax.oss.driver.internal.core.context.InternalDriverContext;
+import com.datastax.oss.driver.internal.core.metadata.schema.events.TypeChangeEvent;
 import com.datastax.oss.driver.internal.core.session.DefaultSession;
 import com.datastax.oss.driver.internal.core.session.RequestProcessor;
 import com.datastax.oss.driver.internal.core.util.concurrent.CompletableFutures;
+import com.datastax.oss.driver.internal.core.util.concurrent.RunOrSchedule;
 import com.datastax.oss.driver.shaded.guava.common.cache.Cache;
 import com.datastax.oss.driver.shaded.guava.common.cache.CacheBuilder;
+import com.datastax.oss.driver.shaded.guava.common.collect.Streams;
+import com.datastax.oss.protocol.internal.ProtocolConstants;
+import io.netty.util.concurrent.EventExecutor;
+import java.util.Map;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionStage;
 import java.util.concurrent.ExecutionException;
 import net.jcip.annotations.ThreadSafe;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 @ThreadSafe
 public class CqlPrepareAsyncProcessor
     implements RequestProcessor<PrepareRequest, CompletionStage<PreparedStatement>> {
 
+  private static final Logger LOG = LoggerFactory.getLogger(CqlPrepareAsyncProcessor.class);
+
   protected final Cache<PrepareRequest, CompletableFuture<PreparedStatement>> cache;
 
-  public CqlPrepareAsyncProcessor() {
-    this(CacheBuilder.newBuilder().weakValues().build());
+  public CqlPrepareAsyncProcessor(DefaultDriverContext context) {
+    this(CacheBuilder.newBuilder().weakValues().build(), context);
   }
 
   protected CqlPrepareAsyncProcessor(
-      Cache<PrepareRequest, CompletableFuture<PreparedStatement>> cache) {
+      Cache<PrepareRequest, CompletableFuture<PreparedStatement>> cache,
+      DefaultDriverContext context) {
+
     this.cache = cache;
+
+    EventExecutor adminExecutor = context.getNettyOptions().adminEventExecutorGroup().next();
+    context
+        .getEventBus()
+        .register(TypeChangeEvent.class, RunOrSchedule.on(adminExecutor, this::onTypeChanged));
+  }
+
+  private static boolean typeMatches(UserDefinedType oldType, ColumnDefinition def) {
+    if (def.getType().getProtocolCode() != ProtocolConstants.DataType.UDT) return false;
+    UserDefinedType defType = (UserDefinedType) def.getType();
+    return defType.equals(oldType);
```

Comment:
Don't we need to recurse into complex types as well? Collection of UDTs, UDT nested into another UDT/tuple, etc.

Reply:
Yes... yes, it does.  You're 💯 correct @olim7t .  I'll expand the check in the type change handler and add some additional cases to the integration test.  I'm glad you caught this!